core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java

```diff
@@ -538,107 +539,103 @@ public void write(String s, byte[] bytes) {
 
-        MethodCreator registerForReflection = file
-                .getMethodCreator("registerForReflection", void.class, Feature.BeforeAnalysisAccess.class)
-                .setModifiers(Modifier.PRIVATE | Modifier.STATIC);
+        MethodCreator registerForReflection = addMethodRegisterForReflection(file);
+        MethodCreator registerClass = addMethodRegisterClass(file);
 
+        int index = 0;
+        MethodCreator currentRegisterClass = null;
         for (Map.Entry<String, ReflectionInfo> entry : reflectiveClasses.entrySet()) {
-            MethodCreator mv = file.getMethodCreator("registerClass" + count++, void.class, Feature.BeforeAnalysisAccess.class);
-            mv.setModifiers(Modifier.PRIVATE | Modifier.STATIC);
-            registerForReflection.invokeStaticMethod(mv.getMethodDescriptor(), registerForReflection.getMethodParam(0));
-
-            TryBlock tc = mv.tryBlock();
-
-            ResultHandle clazz = tc.loadClassFromTCCL(entry.getKey());
-            //we call these methods first, so if they are going to throw an exception it happens before anything has been registered
-            ResultHandle constructors = tc
-                    .invokeVirtualMethod(ofMethod(Class.class, "getDeclaredConstructors", Constructor[].class), clazz);
-            ResultHandle methods = tc.invokeVirtualMethod(ofMethod(Class.class, "getDeclaredMethods", Method[].class), clazz);
-            ResultHandle fields = tc.invokeVirtualMethod(ofMethod(Class.class, "getDeclaredFields", Field[].class), clazz);
-
+            // To avoid getting MethodTooLargeException, the methods only manage up to 100 entries
+            // To avoid getting ClassTooLargeException, the methods need to manage several (100) entries
+            if (index++ % 100 == 0) {
+                if (currentRegisterClass != null) {
+                    currentRegisterClass.returnVoid();
+                }
+                currentRegisterClass = file.getMethodCreator("registerClass" + count++, void.class,
+                        Feature.BeforeAnalysisAccess.class);
+                currentRegisterClass.setModifiers(Modifier.PRIVATE | Modifier.STATIC);
+            }
+            boolean tryBlock = !entry.getValue().weak && (!entry.getValue().constructors && !entry.getValue().ctorSet.isEmpty()
+                    || !entry.getValue().methods && !entry.getValue().methodSet.isEmpty()
+                    || !entry.getValue().fields && !entry.getValue().fieldSet.isEmpty()) || entry.getValue().serialization;
+            final BytecodeCreator creator;
+            ResultHandle clazz = null;
+            if (tryBlock) {
+                TryBlock tc = currentRegisterClass.tryBlock();
+                CatchBlockCreator cc = tc.addCatch(Throwable.class);
+                cc.invokeVirtualMethod(ofMethod(Throwable.class, "printStackTrace", void.class), cc.getCaughtException());
+                creator = tc;
+                clazz = creator.loadClassFromTCCL(entry.getKey());
```

Comment:
Minor: This results in loading the class twice (it always get loaded in `_registerClass`. One way to avoid this would be to make `_registerClass` return the loaded class.